core/sourcehook/sourcehook.h

```diff
@@ -241,29 +273,242 @@ namespace SourceHook
 	*	@brief Pointer to hook manager interface function
 	*
 	*	The hook manager should store hi for later use if store==true. It should then call hi->SetInfo(...) if hi
-	*   is non-null. The hook manager can return 0 for success or a non-zero value if it doesn't want to be used.
+	*	is non-null. The hook manager can return 0 for success or a non-zero value if it doesn't want to be used.
 	*
 	*	@param hi A pointer to IHookManagerInfo
 	*/
 	typedef int (*HookManagerPubFunc)(bool store, IHookManagerInfo *hi);
 
+	/**
+	*	@brief Pointer to hook manager interface function for member and static functions
+	*
+	*	The hook manager should store hi for later use if store==true. It should then call hi->SetInfo(...) if hi
+	*	is non-null. The hook manager can return 0 for success or a non-zero value if it doesn't want to be used.
+	*
+	*	@param hi A pointer to IHookManagerInfo
+	*/
+
+	class IHookManagerMemberFunc
+	{
+	public:
+		virtual ~IHookManagerMemberFunc() = default;
+		virtual int Call(bool store, IHookManagerInfo *hi) const = 0;
+	};
+
+	class HookManagerPubFuncHandler
+	{
+	public:
+		HookManagerPubFuncHandler()
+		 : staticFunc_(nullptr)
+		 , memberFunc_(nullptr)
+		{
+		}
+
+		explicit HookManagerPubFuncHandler(HookManagerPubFunc func)
+		 : staticFunc_(func)
+		 , memberFunc_(nullptr)
+		{
+		}
+
+		explicit HookManagerPubFuncHandler(IHookManagerMemberFunc* funcHandler)
+		 : staticFunc_(nullptr)
+		 , memberFunc_(funcHandler)
+		{
+		}
+
+		explicit HookManagerPubFuncHandler(const HookManagerPubFuncHandler& other)
+		 : staticFunc_(other.staticFunc_)
+		 , memberFunc_(other.memberFunc_)
+		{
+		}
+
+		explicit HookManagerPubFuncHandler(HookManagerPubFuncHandler&& other)
+		 : staticFunc_(other.staticFunc_)
+		 , memberFunc_(other.memberFunc_)
+		{
+			other.staticFunc_ = nullptr;
+			other.memberFunc_ = nullptr;
+		}
+
+		HookManagerPubFuncHandler& operator=(const HookManagerPubFuncHandler& other)
+		{
+			staticFunc_ = other.staticFunc_;
+			memberFunc_ = other.memberFunc_;
+			return *this;
+		}
+
+		HookManagerPubFuncHandler& operator=(HookManagerPubFuncHandler&& other)
+		{
+			if (this == &other)
+				return *this;
+
+			staticFunc_ = other.staticFunc_;
+			memberFunc_ = other.memberFunc_;
+			other.staticFunc_ = nullptr;
+			other.memberFunc_ = nullptr;
+			return *this;
+		}
+
+		HookManagerPubFuncHandler& operator=(HookManagerPubFunc func)
+		{
+			staticFunc_ = func;
+			memberFunc_ = nullptr;
+			return *this;
+		}
+
+		HookManagerPubFuncHandler& operator=(IHookManagerMemberFunc* funcHandler)
+		{
+			staticFunc_ = nullptr;
+			memberFunc_ = funcHandler;
+			return *this;
+		}
+
+		operator bool() const
+		{
+			return (staticFunc_ != nullptr ||
+				memberFunc_ != nullptr);
+		}
+
+		operator HookManagerPubFunc() const
+		{
+			return staticFunc_;
+		}
+
+		operator IHookManagerMemberFunc*() const
+		{
+			return memberFunc_;
+		}
+
+		int operator()(bool store, IHookManagerInfo *hi) const
+		{
+			if(staticFunc_ != nullptr)
```

Comment:
nit: space between "if" and "(".

Reply:
It looks like staticFunc_ and memberFunc_ are mutually exclusive here.

Can we `assert((staticFunc_ && !memberFunc_) || (!staticFunc_ && memberFunc_))`, use `else` instead, and eliminate the default return?